server/service/installer.go

```diff
@@ -0,0 +1,147 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net/http"
+	"os"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+)
+
+type installerRequest struct {
+	EnrollSecret string `url:"enroll_secret"`
+	Kind         string `url:"kind"`
+	Desktop      bool   `query:"desktop,optional"`
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Check if a prebuilt Orbit installer is available
+////////////////////////////////////////////////////////////////////////////////
+
+type checkInstallerResponse struct {
+	Err error `json:"error,omitempty"`
+}
+
+func (r checkInstallerResponse) error() error { return r.Err }
+
+func checkInstallerEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	// Undocumented FLEET_DEMO environment variable, as this endpoint is intended only to be
+	// used in the Fleet Sandbox demo environment.
+	if os.Getenv("FLEET_DEMO") != "1" {
+		err := errors.New("this endpoint only enabled in demo mode")
+		return checkInstallerResponse{Err: err}, nil
+	}
+
+	req := request.(*installerRequest)
+
+	err := svc.CheckInstallerExistence(ctx, fleet.Installer{
+		EnrollSecret: req.EnrollSecret,
+		Kind:         req.Kind,
+		Desktop:      req.Desktop,
+	})
+
+	if err != nil {
+		return checkInstallerResponse{Err: err}, nil
+	}
+
+	return checkInstallerResponse{}, nil
+}
+
+// CheckInstallerExistence checks if an installer exists in the configured storage
+func (svc *Service) CheckInstallerExistence(ctx context.Context, installer fleet.Installer) error {
+	if err := svc.authz.Authorize(ctx, &fleet.EnrollSecret{}, fleet.ActionRead); err != nil {
+		return err
+	}
+
+	if svc.installerStore == nil {
+		return ctxerr.New(ctx, "installer storage has not been configured")
+	}
+
+	_, err := svc.ds.VerifyEnrollSecret(ctx, installer.EnrollSecret)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "cannot find a matching enroll secret")
+	}
+
+	exists, err := svc.installerStore.Exists(ctx, installer)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "checking installer existence")
+	}
+
+	if !exists {
+		return notFoundError{}
+	}
+
+	return nil
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Retrieve an Orbit installer from storage
+////////////////////////////////////////////////////////////////////////////////
+
+type getInstallerResponse struct {
+	Err error `json:"error,omitempty"`
+
+	file []byte // Used in hijackRender for the response
```

Comment:
I believe this could be a `io.ReadCloser` and then on the hijackRender you can use `io.Copy` (to prevent storing the whole package in memory).

Reply:
(And double check if you need to `Close()` after you are done reading the `io.ReadCloser`.)

Reply:
thank you very much for bringing this up. Fixed!